### ISSUE TITLE
Provide access to a parse tree as an alternative to HTML output

### ISSUE THREAD

**Issue by derhuerst** [highlightjs/highlight.js#1086]
I've just had a quick look at hightlight.js and didn't any discussion on this. So forgive me if this is not the right place or has already been discussed.

**I'd like to propose to make highlight.js more generally usable, or at least pull out the core highlighting logic to make it independent from the output format.** Just like the famous [Pygments has formatters](http://pygments.org/docs/formatters/), the HTML generation would be done by a separate module/package.

highlight.js seems to be the de factor standard from syntax highlighting, even among porject not related to HTML/browsers. See [the slap editor](https://github.com/slap-editor/editor-widget/blob/master/lib/highlight/server.js) and [hicat](https://github.com/derhuerst/hicat/blob/master/index.js) for examples: Both use it and later on manually walk/manipulate the HTML generated by highlight.js.

In the world of lightweight do-one-thing NPM packages, it would make sense to split highlight.js into two parts, **one generating an treeish object representation** (possibly streaming), and **the second generating HTML from it**.

**Comment by isagalaev** [comment]
This idea is hanging in the air for a while. It's never been a priority as it lacked real use-cases, but now as you mentioned those two it makes very real sense. Another problem is that in one or two attempts at doing this it was done as a part of bigger (and controversial) refactorings of the whole core library, so it haven't got merged. I will probably get to it myself eventually, but if you'd fancy giving it a try, I could give some guidance.

**Comment by derhuerst** [comment]
> I will probably get to it myself eventually, but if you'd fancy giving it a try, I could give some guidance.

I'm having difficulties in understanding the inner workings of highlight.js, both because there is not a lot of documentation/comments and because the code is written in a really imperative jump-around way.

Also, it seems like HTML generation is really baked into the core of highlight.js.

**Comment by isagalaev** [comment]
Yes, it was deliberately done so in the past to squeeze the code size as much as possible. Currently this "optimization target" is much less desirable and that's why I'm all for refactoring this.

I'll do a short writeup about the architecture of the core to make it more comprehensible.

**Comment by derhuerst** [comment]
> I'll do a short writeup about the architecture of the core to make it more comprehensible.

This sounds great, especially since there seems to be no render-agnostic snytax highlighter in JavaScript land right now.

**Comment by isagalaev** [comment]
> I'll do a short writeup about the architecture of the core to make it more comprehensible.

A status update on this… Before writing anything I decided to first do a pretty significant upgrade to the parser that I was planning to do since a long time ago anyway but a few bugs came up lately bumped the importance of this, and incidentally I've finally came up with an actual plan on how to implement it. And there'd be no point in documenting the current core as it will change immediately after that.

(If interested, the change in question is one I described in [this blog post][1], under the "Complex modes" section.)

[1]: http://softwaremaniacs.org/blog/2014/07/26/highlight-js-what-next/en/

**Comment by dbkaplun** [comment]
Author of slap text editor here. I was going to open a new ticket for this but looks like @derhuerst phrased it very well.

highlight.js has many other uses than highlighting in the browser. It would be great if parsing was decoupled from HTML generation. See also slap-editor/editor-widget#134. We are considering porting our current highlighting implementation using highlight.js, in combination with HTML manipulation with cheerio, to another option that would not require any HTML manipulation such as [lowlight](https://github.com/wooorm/lowlight) (which itself depends on highlight.js but does not seem to perform any HTML manipulation).

Glad to help out in any way if this feature request moves forward!

**Comment by okonet** [comment]
I'm using https://github.com/wooorm/lowlight and writing my RTF formatter ATM so I'd be interested in support for AST out of the box. I think this would allow adding different formatters as pygment does.

**Comment by yyyc514** [comment]
>  Just like the famous Pygments has formatters, the HTML generation would be done by a separate module/package.

I think our HTML stuff has some unique benefits, but if there were two pieces people could always choice to use the one we bundled or replace it with a different once.

I've been digging into the parser a lot, so I'll keep this in mind.

@derhuerst Any idea what that parsed but not HTML format might look like?

**Comment by derhuerst** [comment]
> @derhuerst Any idea what that parsed but not HTML format might look like?

Usually you annotate [Abstract Syntax Trees (ASTs)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) generated by a parser with additional fields, e.g. `highlight.js: {color: '#123123', bold: true}`.

The specific mechanism to add this info of course depends on the specific AST format being used. The most well-known one is [ESTree](https://github.com/estree/estree) used by Mozilla and the [Babel AST format](https://babeljs.io/docs/en/next/babel-parser.html#output) (an extension of ESTree). There's also the [`unist` AST format](https://github.com/syntax-tree/unist#syntax-tree), which tries to support generic trees of markup, e.g. [for HTML](https://github.com/syntax-tree/hast) and [for Markdown](https://github.com/syntax-tree/mdast).

All of these AST formats have dozens or even hundreds of libraries built around them, supporting all kinds of use cases, from parsing to transformations to formatting.

**Comment by yyyc514** [comment]
If we add this support to highlight.js natively, do we kill the whole lowlight project?

**Comment by wooorm** [comment]
Yes, probably!
Although, `highlight.js` main focus is to create an HTML string, whereas `lowlight`s main focus is the syntax tree.
So I imagine highlight’s API to still be mostly the same, with new functions added on top, whereas lowlights internals will change but API stay the same?

**Comment by yyyc514** [comment]
Well sure, we wouldn't change our API.  We'd likely add a new method or two to return some sort of node tree instead of HTML... and maybe some method to turn a node tree into HTML... then we'd just wrap ourselves by glueing the two together with the old api.

**Comment by yyyc514** [comment]
@wooorm Before building your own parser that I assume is derived from ours did you try to first add the functionality to our source?  If so was there a reason you gave up on that approach and went a different way?

**Comment by wooorm** [comment]
I needed hast, and hast was new (my invention). If I recall correctly, people weren't interested in ASTs as much back then so I rolled my own instead of raising an issue requesting my own format to be supported

**Comment by yyyc514** [comment]
Well I meant the core functionality. Once you had a list of parsed tokens you could turn it 8nto any format pretty easily I’d think.

**Comment by wooorm** [comment]
The core functionality is the same for highlight.js and lowlight, except in code style, and in one creating HTML and the other a syntax tree. I think my above comment clarifies that no, I did not raise an issue, and gives reasoning for why not? [Here is a brief history for lowlight](https://github.com/remarkjs/remark-react/issues/14#issuecomment-174311186).

**Comment by yyyc514** [comment]
>The core functionality is the same for highlight.js and lowlight, except in code style, 

Except it's not:

https://github.com/highlightjs/highlight.js/pull/2209
https://github.com/highlightjs/highlight.js/pull/2179
https://github.com/highlightjs/highlight.js/pull/2135

All 3 of these will likely bite you eventually unless you port the same changes to your rewrite... if I were you I wouldn't want to keep up with things like that... that's one of the disadvantages of rewriting vs reusing the core engine.

**Comment by wooorm** [comment]
Yes, there are differences, and for the last almost-four-years I’ve worked on porting them over.

You assume that I want to maintain lowlight like this for years to come. I’d prefer for highlight to support syntax trees. You already asked this and [I responded](https://github.com/highlightjs/highlight.js/issues/1086#issuecomment-542082650):
> Yes, probably! [...] So I imagine highlight’s API to still be mostly the same, with new functions added on top, whereas lowlights internals will change but API stay the same?

**Comment by yyyc514** [comment]
Just making sure you were aware. :-)  I might play around with this idea just to see what the hook in points are... I don' think this would be a difficult thing to do, just need to figure out how to do it nicely.  Our existing system isn't very modular. :)

**Comment by yyyc514** [comment]
Something like this?  Thoughts?

Content:
```
<p class="normal">
<p class="x{{className}}x">
```

Output:
```
<span class="xml"><span class="hljs-tag">&lt;<span class="hljs-name">p</span> <span class="hljs-attr">class</span>=<span class="hljs-string">"normal"</span>&gt;</span>
<span class="hljs-tag">&lt;<span class="hljs-name">p</span> <span class="hljs-attr">class</span>=<span class="hljs-string">"x</span></span></span><span class="hljs-template-variable">{{className}}</span><span class="xml"><span class="hljs-tag"><span class="hljs-string">x"</span>&gt;</span>
</span>
```

Parsetree:

```
{
    "root": true,
    "children": [
        {
            "children": [
                {
                    "tagged": "tag",
                    "children": [
                        "<",
                        {
                            "tagged": "name",
                            "text": "p"
                        },
                        " ",
                        {
                            "tagged": "attr",
                            "text": "class"
                        },
                        "=",
                        {
                            "tagged": "string",
                            "text": "\"normal\""
                        },
                        ">"
                    ]
                },
                "\n",
                {
                    "tagged": "tag",
                    "children": [
                        "<",
                        {
                            "tagged": "name",
                            "text": "p"
                        },
                        " ",
                        {
                            "tagged": "attr",
                            "text": "class"
                        },
                        "=",
                        {
                            "tagged": "string",
                            "text": "\"x"
                        }
                    ]
                }
            ],
            "tagged": "xml",
            "sublanguage": true
        },
        {
            "tagged": "template-variable",
            "text": "{{className}}"
        },
        {
            "children": [
                {
                    "tagged": "tag",
                    "children": [
                        {
                            "tagged": "string",
                            "text": "x\""
                        },
                        ">"
                    ]
                },
                "\n"
            ],
            "tagged": "xml",
            "sublanguage": true
        }
    ]
}

```

**Comment by yyyc514** [comment]
The strange way it splits the tokens across the sublanguage is just how the parser currently works...  Seems wrong to me, but the idea here is that XML is a language "inside Handlebars", not that XML is the primary language and handlebars exists inside that.

The continuations feature is what allows the sublanguage to keep it's context while jumping in and out of handlebars blocks.

**Comment by wooorm** [comment]
on hljs@~9.15, lowlight gives:

```js
{
  relevance: 4,
  language: 'xml',
  value: [
    {
      type: 'element',
      tagName: 'span',
      properties: { className: [ 'hljs-tag' ] },
      children: [
        { type: 'text', value: '<' },
        {
          type: 'element',
          tagName: 'span',
          properties: { className: [ 'hljs-name' ] },
          children: [ { type: 'text', value: 'p' } ]
        },
        { type: 'text', value: ' ' },
        {
          type: 'element',
          tagName: 'span',
          properties: { className: [ 'hljs-attr' ] },
          children: [ { type: 'text', value: 'class' } ]
        },
        { type: 'text', value: '=' },
        {
          type: 'element',
          tagName: 'span',
          properties: { className: [ 'hljs-string' ] },
          children: [ { type: 'text', value: '"normal"' } ]
        },
        { type: 'text', value: '>' }
      ]
    },
    { type: 'text', value: '\n' },
    {
      type: 'element',
      tagName: 'span',
      properties: { className: [ 'hljs-tag' ] },
      children: [
        { type: 'text', value: '<' },
        {
          type: 'element',
          tagName: 'span',
          properties: { className: [ 'hljs-name' ] },
          children: [ { type: 'text', value: 'p' } ]
        },
        { type: 'text', value: ' ' },
        {
          type: 'element',
          tagName: 'span',
          properties: { className: [ 'hljs-attr' ] },
          children: [ { type: 'text', value: 'class' } ]
        },
        { type: 'text', value: '=' },
        {
          type: 'element',
          tagName: 'span',
          properties: { className: [ 'hljs-string' ] },
          children: [ { type: 'text', value: '"x{{className}}x"' } ]
        },
        { type: 'text', value: '>' }
      ]
    }
  ]
}
```

---

It’s a bit verbose, so you can either go two routes:
1. Benefit from the hast ecosystem by exposing hast. [It has a ton of utilities](https://github.com/syntax-tree/hast#list-of-utilities).
2. Go for a simpler AST, such as using strings instead of `type: text`s; using hljs token names instead of elements (`kind: 'attr'`). With 2. you’d get (basically what you have):

```js
{
  relevance: 4,
  language: 'xml',
  value: [
    {
      kind: 'tag',
      children: [
        '<',
        {kind: 'name', children: [ 'p' ]},
        ' ',
        {kind: 'attr', children: [ 'class' ]},
        '=',
        {kind: 'string', children: [ '"normal"' ]},
        '>'
      ]
    },
    '\n',
    {
      kind: 'tag',
      children: [
        '<',
        {kind: 'name', children: [ 'p' ]},
        ' ',
        {kind: 'attr', children: [ 'class' ]},
        '=',
        {kind: 'string', children: [ '"x{{className}}x"' ]},
        '>'
      ]
    }
  ]
}
```

**Comment by yyyc514** [comment]
The  `{{}}` is from handlebars...  yes it's a lot cleaner if you run it as a single language, but I purposely did the more complex route since a lot of complexity is hidden in our sublanguage stuff.  I wonder what does your parser kick out for handlebars (on our latest master branch).  

I see no reason to prefer a complex format.  Someone can write a plug-in rather quickly that re-codes the tree into more complex formats without much difficulty. For example if you preferred your format you could just wrap the new tree and then generate your preferred AST from it. 

I'm talking about doing this just to clean up the internal structure of the parser... it's just serendipitous that it could be user for all these other applications.

**Comment by yyyc514** [comment]
Why do you prefer children/array vs text for text nodes?

**Comment by wooorm** [comment]
> The {{}} is from handlebars... yes it's a lot cleaner if you run it as a single language, but I purposely did the more complex route since a lot of complexity is hidden in our sublanguage stuff. I wonder what does your parser kick out for handlebars (on our latest master branch).

I didn‘t know that! You didn’t mention that your example used handlebars, so I assumed XML. For handlebars the lowlight result looks the same structurally, but has different nodes.
I’m not sure what lowlight spits out on master because it probably has some changes, that I need to backport. Feel free to try.

> Plugins that work "on-top" are trivial because you can just use the global hljs and assign your plugin to it as a function... ie hljs.parseToHast

People in the Node ecosystem seem to like things to be modular and contained, instead of patched on top of other things and globals. Over the years I’ve found that useful as well. So for lowlight I’ll stick with a separate contained package instead of adding a method.

> I'm talking about doing this just to clean up the internal structure of the parser... it's just serendipitous that it could easily help with these other uses cases.

That’s okay! I think (and as others in this thread above) that syntax trees are really interesting and allow for cool things. So I’d argue that both are important.

---

hast is for HTML, and built on top of unist, which is for content and other syntax trees. The goal is different than that of highlight.js. You can do whatever you want, I’m giving two examples.

The reason for having complete nodes instead of just strings is that nodes often have more information, such as positional info. You can read more about [hast in its readme](https://github.com/syntax-tree/hast), and you may find [unifiedjs.com](https://unifiedjs.com) interesting as well!

**Comment by yyyc514** [comment]
> I didn‘t know that! You didn’t mention that your example used handlebars, so I assumed XML. For handlebars the lowlight result looks the same structurally, but has different nodes.

Does it handle sublanguage the same though?  Actually you shouldn't need the new changes since all I used was a boring old handlebars tag...

> People in the Node ecosystem seem to like things to be modular and contained, instead of patched on top of other things and globals. Over the years I’ve found that useful as well. So for lowlight I’ll stick with a separate contained package instead of adding a method.

Well I was suggesting how to do it more as a "plugin" rather than "it's own thing"... like if you wanted to just throw a JS file after Highlight.js loads and 'enhance HLJS'... of course there are a zillion ways to go about it if we exposed some type of low-level API.

> hast is for HTML, and built on top of unist, which is for content and other syntax trees. The goal is different than that of highlight.js. You can do whatever you want, I’m giving two examples.

I was merely saying the best way to go (at first) is likely the simplest... and then let people take that and run with it wherever they want - transforming trees is a pretty simple thing to do. And actually it feels more modular to do it that way also rather than forcing more complexity on us that we need... we build a small simple tree that works best for us internally, and then anyone can take it and transform it however they wish.

> The reason for having complete nodes instead of just strings is that nodes often have more information, such as positional info. 

True, but I'm not sure we're actually equipped to easily provide that type of info anyways at this point.  Food for thought, for sure.  You may have persuaded me to leave strings a a single child node (when they have a kind) though vs making them a whole separate type of key.

**Comment by wooorm** [comment]
> Does it handle sublanguage the same though? Actually you shouldn't need the new changes since all I used was a boring old handlebars tag...

If you want to check out how lowlight does it, I suggest using [runkit](https://npm.runkit.com/lowlight), for example, with this:

```js
var util = require("util")
var lowlight = require("lowlight")

var doc = `<p class="normal">
<p class="x{{className}}x">`

var tree = lowlight.highlight('hbs', doc)

var res = util.inspect(tree, {depth: null})
console.log(res)
```

I get:

```js
{ relevance: 5,
  language: 'hbs',
  value:
   [ { type: 'element',
       tagName: 'span',
       properties: { className: [ 'xml' ] },
       children:
        [ { type: 'element',
            tagName: 'span',
            properties: { className: [ 'hljs-tag' ] },
            children:
             [ { type: 'text', value: '<' },
               { type: 'element',
                 tagName: 'span',
                 properties: { className: [ 'hljs-name' ] },
                 children: [ { type: 'text', value: 'p' } ] },
               { type: 'text', value: ' ' },
               { type: 'element',
                 tagName: 'span',
                 properties: { className: [ 'hljs-attr' ] },
                 children: [ { type: 'text', value: 'class' } ] },
               { type: 'text', value: '=' },
               { type: 'element',
                 tagName: 'span',
                 properties: { className: [ 'hljs-string' ] },
                 children: [ { type: 'text', value: '"normal"' } ] },
               { type: 'text', value: '>' } ] },
          { type: 'text', value: '\n' },
          { type: 'element',
            tagName: 'span',
            properties: { className: [ 'hljs-tag' ] },
            children:
             [ { type: 'text', value: '<' },
               { type: 'element',
                 tagName: 'span',
                 properties: { className: [ 'hljs-name' ] },
                 children: [ { type: 'text', value: 'p' } ] },
               { type: 'text', value: ' ' },
               { type: 'element',
                 tagName: 'span',
                 properties: { className: [ 'hljs-attr' ] },
                 children: [ { type: 'text', value: 'class' } ] },
               { type: 'text', value: '=' },
               { type: 'element',
                 tagName: 'span',
                 properties: { className: [ 'hljs-string' ] },
                 children: [ { type: 'text', value: '"x' } ] } ] } ] },
     { type: 'element',
       tagName: 'span',
       properties: { className: [ 'hljs-template-variable' ] },
       children: [ { type: 'text', value: '{{className}}' } ] },
     { type: 'element',
       tagName: 'span',
       properties: { className: [ 'xml' ] },
       children: [ { type: 'text', value: 'x">' } ] } ] }
```

> I was merely saying the best way to go (at first) is likely the simplest... and then let people take that and run with it wherever they want - transforming trees is a pretty simple thing to do. And actually it feels more modular to do it that way also rather than forcing more complexity on us that we need... we build a small simple tree that works best for us internally, and then anyone can take it and transform it however they wish.

You should do what makes sense for hljs! I do argue that internally a couple of objects instead of strings won't make much of a performance impact, and may even help with readability of code? Whatever floats your boat!
Integrating with a semi-standard like hast does make sense though. One other example is that KaTeX has a similar issue open. So if people start using standard, it’ll benefit a bunch of people.
(But again, whatever works for hljs!)

**Comment by yyyc514** [comment]
> I do argue that internally a couple of objects instead of strings won't make much of a performance impact, and may even help with readability of code? Whatever floats your boat!

Well I'm probably going to join strings, but I might leave them as children.  right now I have a "cleanup" step that wraps all the "only text" children into "text: " nodes, but I'm actually not sure we need that.  I just wrote a quick `walk` and build API:

```js
  class HTMLBuilder {
    constructor() {
      this.out = ""
    }
    addText(text) {
      this.out += escape(text)
    }
    startNode(node) {
      var className = node.kind || node.sublanguage
      if (node.sublanguage || node.kind) {
        this.out += buildSpan(className, "", true, node.sublanguage)
        node.requiresClose = true
      }
    }
    endNode(node) {
      if (node.requiresClose) {
        this.out += spanEndTag
      }
    }
    value() {
      return this.out;
    }
  }
```

If it's really that simple I think I'd prefer to keep it simple and let people add whatever they want on top just by walking the tree. It's possible this is too coupled since I only have the single use case, but seems pretty easy to expand upon later if need be.

**Comment by wooorm** [comment]
Maybe you can investigate how to integrate with document.createElement, React.createElement, Vue..., etc, to see how well it works in other cases?

**Comment by yyyc514** [comment]
That's kind of out of scope though if you want to play around with it and have questions I'm happy to try and answer them... This just generates the raw HTML (which should be pretty easy to hook into most anything).  I was just showing how easy it would be to go from the raw tree to "fully baked" HTML (like we already provide)... that's literally ALL the code someone would have to write to start with the RAW tree and build WHATEVER they wanted.

When I have it a little cleaned up I'll push a branch or PR for someone who'd like to play around with it a little.

In my experience in the past it's been faster to let the browser parser HTML (it's very good at that)... if someone was building an editor or doing selective updates though that's a whole other ballpark and a very custom use case.

**Comment by yyyc514** [comment]
I'm not really sure how to expose it though.  I don't want to lock us into an API super early... for a version one I think we might just expose a `parse` function... that would a return a parseTree...

It might or might not include walk... (or we mark it experimental).... in any case in or out it shows that walking the tree is like 10 lines of code.

**Comment by wooorm** [comment]
> That's kind of out of scope though if you want to play around with it and have questions I'm happy to try and answer them... 

It is the scope of this issue, the title is to decouple from HTML, and allow alternatives. That’s why I think it’s good to think about other formats.

Pygments has different formatters as well as noted earlier, maybe RTF is a good one to try out if you don’t want to into (virtual) DOMs?

> if someone was building an editor or doing selective updates though that's a whole other ballpark and a very custom use case

I think it’s not really a weird case, as it was mentioned in this thread a couple of times over the years.

**Comment by yyyc514** [comment]
Well sure. But only so much time. In this case I think the problems here are pretty well understood conceptually and can be solved abstractly. Any type of alternative renderer needs a simple parse tree to walk, regardless if the target was Elm or a printed PDF. The goal (in my mind) is to build something people could reasonably build on top of. I can throw something together and push a beta. Someone else in each of those domains needs to come out of the woodwork to test and see how well it works for them.for example you could test it with your thing to see how easy it is to work with. Etc. 

And even if it’s not helpful at all I think it’ll still make the code easier to understand because it woks it’s two huge concerns that are currently very intermingled but shouldn’t be. 

> Pygments has different formatters as well as noted earlier, maybe RTF is a good one to try out if you don’t want to into (virtual) DOMs?

Not really a matter of want. And again a lot of this is hinged on fact that at the most all you’d need is a translation layer between the raw parse tree and a more cooked parse tree that RTF or whatever desired. It shouldn’t be much more code than the HMTL builder unless I’m really missing something. 

> I think it’s not really a weird case, as it was mentioned in this thread a couple of times over the years.

Other maintainers have said pretty clearly we can’t support that though. We’re designed to render static code blocks, not serve as the basis for a live code editor. Not saying it’s impossible or anything just that you’re really pushing the limits and we can’t be responsible if you blow yourself up or it doesn’t work as well as you would hope.

**Comment by yyyc514** [comment]
Some type of auto diving shadow Dom thing might be cool thought for the editor people. But I think you could actually do that already with just the HTML output too so I’m not sure this would be a huge improvement in that category.

**Comment by okonet** [comment]
> maybe RTF is a good one to try out if you don’t want to into (virtual) DOMs?

I actually implemented RTF formatter using `highlight.js` and `ace.js` and could share that (it's OSS). Check out https://github.com/okonet/codestage/tree/master/lib it's decoupled from the implementation but relies on HTML now. 

The worst and hardest problem when using HTML output as the input is to match CSS selectors. I had to pull in JSDOM just to do that job since the actual implementation is quite complex.

If I'd had an AST of the highlighted source, it would be super straight-forward to implement. The actual RTF renderer is just this: https://github.com/okonet/codestage/blob/5cf13e2ba80188b0bfbead8f6d0626e4839c0afa/lib/src/index.js#L76-L108

**Comment by yyyc514** [comment]
> The worst and hardest problem when using HTML output as the input is to match CSS selectors.

Aren't they just simple strings?  I mean you'd have to parse the HTML again to get nodes to walk, but once you had those it's just "hljs-variable" or such, no?  In the AST they'd be "variable"... same thing without the prefix.

**Comment by okonet** [comment]
Yes, but I don’t need nodes, I need computed styles of the node to transform it to RTF instructions. Matching nodes and getting computed styles are quite hard if you’re running outside of the browser. That’s why the most reliable way was to stick with JSDOM. The simpler approach worked for simple themes (one selector per node), but not for more complex ones. Also some themes are relying on CSS cascade so simple matching using 2 ASTs (html and css) didn’t work 🤷🏻‍♂️

**Comment by yyyc514** [comment]
Oh. I get it. You have to write a render engine too. That’s probably impossible to do fully without parsing CSS fully and all it’s insanity.

**Comment by yyyc514** [comment]
I don’t see how the AST we’d build here would help you then. It would only include classes. You’d still have to figure out how CSS would apply those classes in the real world, no?

**Comment by okonet** [comment]
Maybe you’re right. I’d hope the AST would not include classes but rather semantics of the tokens. The renderer would then either apply class names to Dom nodes or use a different technique like RTF instructions. Having class names in the AST won’t help, yes.

**Comment by okonet** [comment]
JSDOM is the best you can get but I’m wondering if relying on it is the right approach. To me, since this project does parsing and “knows” what nodes are representing, why should it rely on DOM at all. This is a rendering target to me. That said, I realize this would probably apply some limitations on current theming approach and will definitely be a breaking change. But I’d be glad to assist in any way with the RTF renderer or anything else if that’s interesting

**Comment by wooorm** [comment]
@okonet You seem to be interesting in the link between CSS and the HTML nodes, right? Which I don’t think is solved by this issue/PR indeed: even when tokens here wouldn’t represent the DOM, they still aren’t “linked” to their styles

That’s a hard problem! 🤷‍♂️

**Comment by yyyc514** [comment]
>  I’d hope the AST would not include classes but rather semantics of the tokens. 

All we have are class names, that's all the semantics we have - the styles from the styling.  If you had a RTF or something you'd have to define your own "style" and then decide what each of the classes represented - we do have a fixed and documented list.

>  To me, since this project does parsing and “knows” what nodes are representing, why should it rely on DOM at all.

We don't, we only use it for some of the tests to sort of make sure the "browser" payload seems to work in a browser like environment.  I don't think I was suggesting it in any other usage.  We're pretty happy with having 0 requirements for the client-side I think.

You wouldn't need it if you just consumed the parse-tree directly.

> But I’d be glad to assist in any way with the RTF renderer or anything else if that’s interesting

Sounds like a great plug-in idea to me.

> Which I don’t think is solved by this issue/PR indeed: even when tokens here wouldn’t represent the DOM, they still aren’t “linked” to their styles

Yeah, you'd just have `class`, not actual styling... so you'd have one or two "paper" themes (not sure what other targets you'd use RTF fun, but could be my lack of knowledge here)...

**Comment by yyyc514** [comment]
Latest update:

https://github.com/highlightjs/highlight.js/pull/2404

**Comment by yyyc514** [comment]
This is merged into master now.  I wouldn't call this a 100% public API just yet but you can get the parse tree object now with:

```
result.emitter.root
```

Check out `token_tree.js`.  You can also replace the whole emitter with `configure({__emitter: CustomEmitter })` though that's definitely flagged as beta. It seems a reasonable extension point to me though, hence trying it out.

Very curious to know if that helps any of you. CC @wooorm

**Comment by wooorm** [comment]
Awesome, I'll check it out!
While its beta, can we somehow have a contract that it would only change in minor / major versions, and not in patches?

**Comment by wooorm** [comment]
It works well, it’s a bit hard to figure out without docs, and https://github.com/highlightjs/highlight.js/issues/2522 and https://github.com/highlightjs/highlight.js/issues/2523 are blockers for upgrading lowlight. Thanks @yyyc514.

What I’m missing is a way to extend highlight.js inside lowlight with options, while not affecting highlight.js for other consumers (e.g., someone expecting a string). Right now I must get the current configuration, then configure, and the finally configure the previous configuration again.

**Comment by yyyc514** [comment]
>  it’s a bit hard to figure out without docs

Well docs are always better, sure, but I did try to write the code and APIs clearly... if you have any specific thoughts on that I'd love to hear them.

**Comment by yyyc514** [comment]
```
const instance = HLJS({});
instance.newInstance = HLJS.bind(null, {});

// export an "instance" of the highlighter
export default instance;
```

Would this get the job done?

**Comment by yyyc514** [comment]
OR if you're building from source just pull in the RAW ES6 modules and then call HLJS() yourself to get your own instance... I'm not sure if using the built version of the libraries is a MUST for what you're doing or not.

**Comment by wooorm** [comment]
I’ll try `hljs.newInstance`! ✨

> I think all you'd need to understand is the emitter API (it's a few methods) and then the tree structure itself... see it's toJSON method. You either want to be an emitter yourself or you want to walk the tree afterwards.

For the emitters, it seems like the code is currently divided into three acts, HTML renderer, Token Tree, and Token Tree Emitter, I’m not sure I follow why so many are needed? What’s the case for a toHTML, toJSON, and walking? Could emitters build a single `value`, which could be anything, that is returned?

Until I got hit by the infinite loop, I was able to construct the trees I need with (would clean it a bit if everything is working and when I know what’s required and what’s optional):

```js
Emitter.prototype.add = noop
Emitter.prototype.addText = text
Emitter.prototype.addKeyword = addKeyword
Emitter.prototype.addSublanguage = addSublanguage
Emitter.prototype.openNode = open
Emitter.prototype.closeNode = close
Emitter.prototype.closeAllNodes = noop
Emitter.prototype.finalize = noop
Emitter.prototype.toHTML = toHtmlNoop
Emitter.prototype.toJSON = noop
Emitter.prototype.walk = noop

function Emitter(options) {
  this.options = options
  this.rootNode = {children: []}
  this.stack = [this.rootNode]
}

function addKeyword(value, name) {
  this.openNode(name)
  this.addText(value)
  this.closeNode()
}

function addSublanguage(other, name) {
  var stack = this.stack
  var current = stack[stack.length - 1]
  var results = other.rootNode.children

  current.children = current.children.concat(
    name ?
      {type: 'element', tagName: 'span', properties: {className: [name]}, children: results} :
      results
  )
}

function text(value) {
  if (value === '') return
  var stack = this.stack
  var current = stack[stack.length - 1]
  var tail = current.children[current.children.length - 1]
  if (tail && tail.type === 'text') tail.value += value
  else current.children.push({type: 'text', value: value})
}

function open(name) {
  var className = this.options.classPrefix + name
  var stack = this.stack
  var current = stack[stack.length - 1]
  var child = {type: 'element', tagName: 'span', properties: {className: [className]}, children: []}
  current.children.push(child)
  stack.push(child)
}

function close() {
  this.stack.pop()
}

function toHtmlNoop() {
  return ''
}

function noop() {}
```

I’m not sure the three layers are needed internally, why not a single HTMLRenderer?

---

The second question/suggestion is that I’m missing an abstract version of highlight / highlightAuto: something that could be used in hljs for the HTML results, or for others to get different results, instead of having to get `result.emitter.rootNode`. Right now I’m overwriting toHTML to not build a string, which may be hacky? I don’t know! 🤷‍♂️

**Comment by yyyc514** [comment]
> I’ll try hljs.newInstance!

Well it doesn't exist yet, lol... I was asking if that's really all you needed... just your own instance.

**Comment by yyyc514** [comment]
> For the emitters, it seems like the code is currently divided into three acts, HTML renderer, Token Tree, and Token Tree Emitter, I’m not sure I follow why so many are needed? What’s the case for a toHTML, toJSON, and walking? Could emitters build a single value, which could be anything, that is returned?

**The TokenTree is a purely conceptual data storage.**  Is just has tree stuff, simple node management, walking the tree.   `toJSON` is for debugging use. It's just data and tree utility functions.  IE, it could be used stand-alone.

**TokenTreeEmitter is the emitter - this is the component/API that HLJS needs.** It has documentation (in the form of a list) for all you need to support API-wise to be a valid emitter. The TokenTree emitter merely builds a token tree, nothing else. Perhaps it should have embedded not extended the TokenTree...

There are 3 things because there are 3 concerns here: (<https://en.wikipedia.org/wiki/Separation_of_concerns>)

- The back-end data model
- The emitter itself
- HTML generation

Originally there wasn't even a `toHTML` but I added it to make it easier for people who wanted to write simpler emitters.  So the concerns are more cleanly separated, but yes there is a bit more code...

> Could emitters build a single value, which could be anything, that is returned?

**Emitters can do anything they want as long as they support the API.**  A pure string emitter that didn't built a tree would be super simple. You'd just be appending HTML and building a long string...

Your emitter shouldn't need to implement `walk`.

> The second question/suggestion is that I’m missing an abstract version of highlight / highlightAuto: something that could be used in hljs for the HTML results, or for others to get different results, instead of having to get result.emitter.rootNode.

Why can't you just use options to set your own `__emitter`?  You don't have to use the token tree unless you find it valuable.

Your code would probably be shorter if you DID use the token tree though and just call walk and pass it a builder...

**Comment by yyyc514** [comment]
If the existing tree is "suitable" for your purposes you'd only need to write a builder:

```
  static _walk(builder, node) {
    if (typeof node === "string") {
      builder.addText(node);
    } else if (node.children) {
      builder.openNode(node);
      node.children.forEach((child) => this._walk(builder, child));
      builder.closeNode(node);
    }
    return builder;
  }

```

Builder only needs `addText`, and `openNode`, and `closeNode`.  A HTML builder would be 5-15 LOC I'd guess.

```rb
# psudocode
text = ""
addText -> text += txt
openNode -> text += '<span class="arg">'
closeNode ->  text += '</span>'
```

**Comment by yyyc514** [comment]
> that I’m missing an abstract version of highlight / highlightAuto: something that could be used in hljs for the HTML results

Yeah if you don't care about HTML at all you will still need an emitter that stubs `toHTML` just to satisfy the API, but as you see it can return `""`... Then your code could use whatever API you exposed to get the "real" data you wanted.

**Comment by yyyc514** [comment]
I bet your code would be 25-50% less if you rewrote it as a builder vs emitter.  So much of your complexity is managing your stack/tree stuff. Of course doing that you'd still pay the CPU cost to covert to HTML, although if you had your own instance you could just "patch" the built-in emitter to make `toHTML` a NOOP.

**Comment by yyyc514** [comment]
`
 if (tail && tail.type === 'text') tail.value += value
`

Although perhaps you're doing collapsing too... so then you might not be able to use the tree as-is, because I don't think for the core library collapsing is super important, so that function may disappear completely in the future.

**Comment by yyyc514** [comment]
From "outside" the library your main choices are:

- hack emitter
- replace emitter
- use the token tree directly (via rootNode, etc)
- `walk` the token tree

If you pulled in the raw modules you could use the individual components as building blocks - that's why they are abstracted as such... It'd be trivial to subclass TokenTreeEmitter for example to add a `toPDF()`... and then add a PDFRenderer... The renderer doesn't have to concern itself with the tree, or the data storage... it just has to respond to open/close/more text events.

Actually I guess that's an example of something you could do WITHOUT the raw modules too. :)

**Comment by yyyc514** [comment]
You shouldn't need toJSON either... if my comments are wrong, let me know:

```

  Minimal interface:

  - addKeyword(text, kind)
  - addText(text)
  - addSublanguage(emitter, subLanguageName)
  - finalize()
  - openNode(kind)
  - closeNode()
  - closeAllNodes()
  - toHTML()
```

**Comment by yyyc514** [comment]
I was originally going to also provide a text emitter (which would be nice to look at for a second example), but then we'd need to find a way to double-run the test suite for both emitters and it might limit our flexibility in the future... and the idea of including a whole class that no one will ever use is a little silly.

**Comment by wooorm** [comment]
> hljs.newInstance

Is there a reason why HLJS isn’t a normal class or prototypal constructor, of
which an actual instance is exposed?

https://github.com/highlightjs/highlight.js/blob/e8624ccb65cfd859a0273b58ec3897afd16a5498/src/highlight.js#L723

If an instance is exposed, the constructor can be accessed from it with `HLJS = hljs.constructor; myInstance = new HLJS()`

---

> You probably want 2 or 4 and it looks like 2 is what you started with.

I think 4 options are too many. It makes it confusing for me which one to pick. Internally, you’re now building a token tree which you’re not using, that’s unneeded memory! And the API surface which has to be maintained also grows a lot with toJSON / toHTML / walk.

I would suggest using a single interface internally, in highlight.js, that essentially works as how it’s implemented now, but merged into one thing.

The API then is the interface: not a subclass or one of 4 options, but an object that can handle open/close/text/etc.

I would propose a smaller interface:
- constructor given `options`
- `value` property which is the result returned by `highlight()`
- `text` method handles raw text
- `open` method, given a token name, opens a token
- `close` method closes a token
- `sublanguage` method, given the `value` from another builder instance and a language name.

(I don’t think keyword needs to be handled differently, it’s open/text/close, right?)

The internal HTML builder/adapter/emitter could then look like:

```js
Builder.prototype.text = text
Builder.prototype.open = open
Builder.prototype.close = close
Builder.prototype.sublanguage = sublanguage

function Builder(options) {
  this.options = options
  this.value = ''
}

function text(value) {
  this.value += encode(value)
}

function open(name) {
  this.value += '<span class=' + encode(this.options.classPrefix + name) +  '>'
}

function close() {
  this.value += '</span>'
}

function sublanguage(value, name) {
  if (name) this.value += '<span class=' + encode(name) +  '>'
  this.value += value
  if (name) this.value += '</span>'
}
```

Whereas the lowlight one would look like:

```js
Builder.prototype.text = text
Builder.prototype.open = open
Builder.prototype.close = close
Builder.prototype.sublanguage = sublanguage

function Builder(options) {
  this.options = options
  this.value = {type: 'root', children: []}
  this.stack = [this.value]
}

function text(value) {
  if (value === '') return
  var stack = this.stack
  var current = stack[stack.length - 1]
  var tail = current.children[current.children.length - 1]
  if (tail && tail.type === 'text') tail.value += value
  else current.children.push({type: 'text', value: value})
}

function open(name) {
  var className = this.options.classPrefix + name
  var stack = this.stack
  var current = stack[stack.length - 1]
  var child = {type: 'element', tagName: 'span', properties: {className: [className]}, children: []}
  current.children.push(child)
  stack.push(child)
}

function close() {
  this.stack.pop()
}

function sublanguage(subtree, name) {
  var stack = this.stack
  var current = stack[stack.length - 1]

  current.children = current.children.concat(
    name ?
      {type: 'element', tagName: 'span', properties: {className: [name]}, children: subtree.children} :
      results
  )
}
```

I believe this interface has a right amount of freedom compared to extensibility and being straightforward and relatively simple.

It’s a bit more like existing tools with similar interfaces, such as [sax](https://github.com/isaacs/sax-js) or [parse5](https://github.com/inikulin/parse5/blob/master/packages/parse5/docs/tree-adapter/interface.md)

(another idea for a builder/adapter/emitter is to generate ANSI colors to print in terminals.)

**Comment by yyyc514** [comment]
> Is there a reason why HLJS isn’t a normal class or prototypal constructor, of
which an actual instance is exposed?

No idea, before my time.  The old module stuff requires you to pass in an object that gets "enhanced".  If I was started from scratch it'd probably be a ES6 class.

> I think 4 options are too many. It makes it confusing for me which one to pick.

I didn't say they were all great options.  I'm speaking to you as someone hacking the core library to do unintended things.  If you just want the happy path it's 2 or 4, and there are different reasons you'd choose one or the other - depending on what your goals were.  

> Internally, you’re now building a token tree which you’re not using, that’s unneeded memory! 

In tests it wasn't significant and some tests even showed it was faster than before.

> And the API surface which has to be maintained also grows a lot with toJSON / toHTML / walk.

`toHTML` is a one liner.  `toJSON` is debugging.  `walk` is necessary if we have a tree, it's just a matter of where you put the code, I felt a tree should know how to walk itself.

Again these aren't things a 3rd party needs to implement - only use - if they need them.

> I would suggest using a single interface internally, in highlight.js, that essentially works as how it’s implemented now, but merged into one thing.

There is a single interface, it's the "Emitter" interface (in a TypeScript sense).  You implement that and you're done.  The fact that I've chosen to implement it with 3 classes is an implementation detail.  Someone writing their own can make a different choice.

Now let me read your actual proposal. :-)

**Comment by yyyc514** [comment]
> I would propose a smaller interface:

You essentially describe almost exactly the same interface we already have, just you renamed a few things.  So lets discuss the differences:

*addKeyword*

This is a convenience method for the parser.  Inside the highlighter a keyword is a very different thing than a block and I just want the parser to focus on parsing, not HTML representation - so it just tells the emitter "hey open a new block" or "hey i have a keyword".  It's the emitters responsibility to know what to do about that - not the parser.  Separation of concerns.

Your'e assuming a keyword is just open/keyword/close.  I think that is based on some strong assumptions.  Perhaps someone wants an emitter that can turn off special treatment of keywords.  Now we don't have those open/close tags at all.  **Keyword is a separate concept, so it has a separate method.**

The new parser knows LESS about the final output format, and that's a good thing. 

*sublanguage*

It really needs both the "content" and the language so that it can scope the new node/class/whatever properly... hence it takes 2 properties.  And it takes an emitter for flexibility.  So if you have a string based emitter, you can indeed call `emitter.value` and just append that to your parents current value, and you're all done.

I suppose you MIGHT imagine that the underlying code was returning it's own wrapper (as a pure string), but I think that's also a abstraction violation IMHO.  The way sublanguages currently are designed they don't need to know anything about the parent - so the most "separate" implementation is one where the sublanguage only returns a structure representing it's highlighted code.

If that needs to be wrapped inside a `<div class="embedded-javascript"></div>` that's a concern of the parent, not the sublanguage rendering.

I think originally one emitter passed raw data to another, but I felt that wasn't encapsulated enough so now `sublanguage` takes the whole emitter and then uses it's API to retrieve the data.

*closeAllNodes* and *finalize*

Again, convenience methods for the parser, that wants to focus on parsing concepts not output rendering.  Finalize is kind of a "covers all bases".  We could also provide some type of base class that had a default NOOP implementation if that was a real concern.

**Comment by yyyc514** [comment]
> (I don’t think keyword needs to be handled differently, it’s open/text/close, right?)

If we had an abstract emitter class it MIGHT perhaps be reasonable to have a default implementation of addKeyword that did this... hard to say without knowing how people are going to use this.

**closeAllNodes**

I will give this some thought.  I found it a big win though to take this responsibility away from the parser and make it a responsibility of the emitter.  The parser naturally works and thinks in trees.  The grammars are trees, and the outputs most natural data representation is a tree. And with a tree of course you have no tags to worry about close. 

If you're 20 levels deep when you finish, then you're still done.  No need to close anything.   **That's an HTML/rendering concern.**

**Comment by yyyc514** [comment]
And again, all this is considered private API and might stay so... it's mostly to make advanced integrations with stuff like lowlight possible and MUCH simpler.  Instead of forking our whole library like you used to now you should be able to do the same thing with 50-100 lines of code.

**Comment by wooorm** [comment]
> And again, all this is considered private API and might stay so... 

This issue is about creating a public API for that. My suggestions aren‘t for how things work internally, but for what is exposed

**Comment by yyyc514** [comment]
So question to you: Can you not build lowlight on the builder API alone?

**Comment by yyyc514** [comment]
The layering is about how low-level you want to be.  The fastest way to get output is a Builder, that would be what I'd steer most non-advanced implementers to... the most complex and customizable way is an Emitter.  Builder is trivially simple, Emitter requires more effort on the part of the implementer.

**Comment by wooorm** [comment]
I was able to get up to a point, but there are/were other issues which I raised in other issues.
I don’t want to update to a private API though. My above comments are feedback on API design that would solve this issue.

> If we made an API public public we might start with the tiny builder API... so we'd support:

You may do that! My suggestion is to support the bare minimum that we need now, in a low-level way that should be powerful enough to do most things. I posted HTML and hast making above, I can see how [emphasize](https://github.com/wooorm/emphasize) could also be made to work for ANSI coloring. I don’t know about PDFs so I can’t make suggestions for that.

Re adapter / builder / emitter: I didn’t study CS; I don’t understand the subtleties you intend between these terms.

**Comment by yyyc514** [comment]
> Re adapter / builder / emitter: I didn’t study CS; I don’t understand the subtleties you intend between these terms.

No subtly there, those terms are mostly made up (they may have some meanings I dunno, I did try to give them meaningful names) - but it's just what I decided to call them.  The separation of concerns and different layers of abstraction are the real things to take away from what I'm talking about.

> I was able to get up to a point, but there are/were other issues which I raised in other issues.

Registration of aliases and getting a new instance are solvable - no objections there. :-)  Though I worry about regular users making instances they don't need, but I guess that's a documentation issue?

> I don’t want to update to a private API though.

**I'd think a semi-private API is far better than forking the whole project.**  You update to the semi-private API and then pin to the latest compatible version of HLJS.  Even if it remains private there are no huge plans to rejigger this stuff all the time.  And now updating 90% of the time is just a matter of bumping the version and running your test suite. No more porting over deep internal parser bug fixes/changes.

That sounds worlds better than what you are doing now - treating the whole codebase as a HUGE private api. :-)

**Comment by yyyc514** [comment]
Now we have a chicken in egg problem because I don't feel these things are best designed/used in the dark.  I'd want a few people to actually **use** the system before we ever declared it a public API and locked ourselves into it for at least a major revision cycle and perhaps MUCH longer.

So I think the cycle has to start with advanced integrators actual using the new stuff in the real world - even if it's not blessed for general public usage and not 100% API stable (from one version to the next).  That's what NPM version pinning is more IMHO.

**Comment by yyyc514** [comment]
But even if no one ever uses it I still think the internals of the parser have been much improved by these changes... EVEN if we dropped the token tree in the future and went back to pure strings as of right I'd still use the same API for the emitter interface. (I use interface here in the TypeScript/API sense).

**Comment by yyyc514** [comment]
So question to you: Can you not build lowlight on the builder/walk API alone?

**Comment by yyyc514** [comment]
Also there are stability/debugging issues as well related to how these concerns are separated.  The plain reading of this issues title (I didn't go back and read the whole thread) was requesting a tree output alternative to string output.

That's what `Emitter#walk` provides. **It does so fully separately from the parsing task.**  It's impossible for a broken Builder to break Highlight.js itself.  Say you had a huge spike of memory or a freeze... it's very easy to draw the line and say:

- Highlight.js is broken
- Your builder is broken

Vs an emitter, which is running in parallel with the parsing process.  It'd be very easy to turn the parse tree into JSON and pass it around, render it later, etc... so the idea of writing an Emitter (that the parser is talking to real-time) vs writing a Builder (that the Tree walker can use after the fact to generate another form of presentation) are different conceptually also - not just in their complexity of implementation.

**Comment by wooorm** [comment]
hljs now has an `__emitter` option. I assumed that was the way to go. I did that, and it mostly works. [You asked for my thoughts](https://github.com/highlightjs/highlight.js/issues/1086#issuecomment-593003028), [I gave them](https://github.com/highlightjs/highlight.js/issues/1086#issuecomment-623072456) based on my existing experience buildings parsers and maintaining hundreds of open source projects.

For context: 53% of hljs downloads on npm come from lowlight, it’s used a lot, I don’t want to provide a potentially bad user experience by depending on private APIs. And I think if lowlight would implement this api, “I'd want a few people to actually use the system before we ever declared” is pretty much solved.

You are the maintainer of hljs and you are free to implement these features the way you want!

**Comment by yyyc514** [comment]
If I've offended I have not meant to. I'm just discussing and explaining my thinking.

>  “I'd want a few people to actually use the system before we ever declared” is pretty much solved.

I meant use/integrate with the API - I don't mean end users, I mean integrators.  I assume before and after the end user experience for your library will remain 100% unchanged - assuming the new functionality on our end actually works, so those really aren't useful data points.  _It's not the end user experience of your library users we're talking about, it's the integrator experience - ie, your experience._  

And at this point you and I are merely two individual data points as having implemented working Emitters. :-) If you can I still urge you to try writing a builder and see how that feels since you're working at a much higher level. It would be helpful to have feedback on that.

> hljs now has an __emitter option. I assumed that was the way to go. I did that, and it mostly works.

This is the problem with features without docs - and truly sorry if I steered ya in the wrong direction originally. :-)  It's a fine way to go, but I just don't think it was clear that _you're working at the lowest-possible level available_ (short of forking the library).  So the API is indeed a bit more complex at that level.  The builder API is a higher level API and therefore much simpler.

If I'd had the hindsight of this whole discussion I may have suggested the builder from the get-go, but hindsight is 20/20. :-)

**Comment by yyyc514** [comment]
> I don’t want to provide a potentially bad user experience by depending on private APIs.

Do you publish on NPM?  Can you not pin your library against a "known good" version of ours? I guess I fail to see how forking the whole library is better/easier than pinning to a specific version of HLJS that provides a specific "known good" API - whether private or not.  

Again, isn't forking a whole project kind of the "worse case" scenario - in that case you're treating the whole project as a giant private API... if we change a ton of stuff, _you're "stuck" on the prior version until you manually update._

How is that any different than using a semi-private API - other than the fact that the amount of code you have to maintain is now reduced by an order of magnitude or two?  If we change a ton of stuff *in general*, you now get all those changes for free. If we change a ton of stuff specifically related to the integration API... _you're "stuck" on the prior version until you manually update._

Seems like option 2 is better than option 1, or at least no worse... no? Am I missing something? What is the big advantage of "vendoring it" (having a static/forked copy to integrate with) vs having a static versioned copy via npm?

I'm not sure the goal here was ever to make a 100% public API that we have to support to the same level of stability that we support our other public APIs.  My goal was to simplify the parser, separate the concerns better, and while doing so make it so that integrators (like yourself) wouldn't necessarily have to maintain a whole custom copy of the library just to get between the `parser <=> render` process.

---

> based on my existing experience buildings parsers and maintaining hundreds of open source projects.

Where do you find the time? :-)  Then surely you understand my hesitation to add and document a public and stable API for a tiny, tiny subset or our users (advanced integrators) and then be locked into that API for a year (or longer).  A year is the pace I currently imagine for major releases where we allow ourselves to truly "break" things.

**Comment by yyyc514** [comment]
It's entirely possible you know more about parsers than me if you maintain a lot of them - I'm just doing the best I can with the knowledge and experience I have. :-)  I have been programming for professionally for 25+ years, so hopefully that's worth something... :-)

> My suggestions aren‘t for how things work internally, but for what is exposed

So perhaps the confusion is over "what do we expose".  I see emitter as a lower-level implementation concern.  It's the interface the parser needs so that it can focus on parsing and not presentation.  I made it replaceable because I feel that's good design (clean separation) plus at the time I imagined we might have to ALSO ship a pure text emitter in additional to the tree emitter... until I did a few tests and the memory/CPU didn't seem to be all that different with the tree version.

A side benefit of that (for someone like yourself) is that now **instead of maintaining a port of the whole project** - you can just track the emitter/builder API - even if the internals aren't 100% API stable.  So when we release a new version you now rewrite 100 lines instead of potentially having to rewrite the whole thing. :-)

I'm not interested in changing this API all the time for no good reason.  I expect it to be pretty stable, but that's different than making a promise to everyone that it's 100% stable.

**Comment by wooorm** [comment]
> If I've offended I have not meant to. I'm just discussing and explaining my thinking.

You do communicate a lot, which is mostly a great thing, but it’s many messages with many edits, and for me it’s hard to keep up! I feel like I’m out of depth (I don’t know, or necessarily want to know, much about highlightjs or how it is maintained); I don’t want my opinion to hold too much weight. My way of dealing with many messages is going in reverse and sending short responses back to keep the conversation manageable for me.

---

> Do you publish on NPM? Can you not pin your library against a "known good" version of ours? I guess I fail to see how forking the whole library is better/easier than pinning to a specific version of HLJS that provides a specific "known good" API - whether private or not.

The fork is from 4+ years ago. We’re already pinning hljs to feature releases, because over the last 4 years it hasn’t adhered to semver, where breaking changes occurred in feature released. I’m personally strongly against lockfiles and completely locking versions.
Yes, lowlight works with the new emitter. I want to release that. Or with (a combination of) my suggestions. I don’t care. But I do want it to be public and good API. What _public_ and _good_ means is up to you.

---

> How is that any different than using a semi-private API - other than the fact that the amount of code you have to maintain is now reduced by an order of magnitude or two? If we change a ton of stuff in general, you now get all those changes for free. If we change a ton of stuff specifically related to the integration API... you're "stuck" on the prior version until you manually update.

Other that initially building lowlight, I have probably spent on 20 hours over 4 years on the 18 releases and it’s been very stable. Pretty sure I’m about to hit those hours discussing this issue. I want a working version that follows a contract where hljs provides something that lowlight can safely work with: it could look like anything, such as how it currently is (but public) or maybe a bit more simple

---

> I'm not sure the goal here was ever to make a 100% public API that we have to support to the same level of stability that we support our other public APIs. My goal was to simplify the parser, separate the concerns better, and while doing so make it so that integrators (like yourself) wouldn't necessarily have to maintain a whole custom copy of the library just to get between the parser <=> render process.


_**ANYTHING**_ you expose, is public. This emitter is public. The `.emitter.rootNode` is public. hljs 10 broke lowlight, which was working on public APIs for years, and is now public! Public doesn’t mean it never changes.

With the rest of the paragraph: I understand! And I’m :+1: with whatever you decide to implement.

---

You are free to create a tree! That’s what was originally asked, and what you delivered. But the original question was a tree instead of HTML. We now have both. And there are 4 ways to integrate. For a tree, you could go with something that  is well defined and highly used. Such as hast, but it was decided to go with a tree. That means that *anyone* now has to adapt the custom hljs tree to what they want. Be it HTML or hast or ANSI or PDF. And you’re doing both a tree and HTML!

I think a simpler solution would be: return whatever is returned by that `__emitter`. Make the emitter powerful enough for hljs HTML, lowlight hast, or ANSI, but support only [4 methods](https://github.com/highlightjs/highlight.js/issues/1086#issuecomment-623072456) in the API. And use the emitter internally by building that HTML so it’s 100% tested.

What I suggest is less public api surface than currently in v10.

P.S. lowlight (and thus hljs) is used in many static site generators, such as Gatsby or Next.js, and it’s running on some blogs with thousands of articles: performance is really important, bazillions of lines are being highlighted

**Comment by yyyc514** [comment]
> You do communicate a lot, which is mostly a great thing, but it’s many messages with many edits, and for me it’s hard to keep up! 

Blessing and a curse. :)  One of your messages just sounded a bit frustrated so I was trying to defuse that if possible.  

> I don’t want my opinion to hold too much weight. 

No worries. I try not to let any single voice hold too much weight. :)

> The fork is from 4+ years ago. We’re already pinning hljs to feature releases, because over the last 4 years it hasn’t adhered to semver, where breaking changes occurred in feature released.

If we do that now, let me know, but we're trying not to.  Though technically it might depend on what you consider a "breaking change".  When you fork the library and have to hand or hand merge things back in I'd imagine far too many things are "breaking changes". On the other hand technically no releases actually break you since you're a fork... :-)

> I’m personally strongly against lockfiles and completely locking versions.

I can understand that I suppose.  I was suggested it as a way to "hedge your bets".

> Other that initially building lowlight, I have probably spent on 20 hours over 4 years on the 18 releases and it’s been very stable.

Interesting and impressive, I'd have guessed more.

> Yes, lowlight works with the new emitter. I want to release that. Or with (a combination of) my suggestions. I don’t care. But I do want it to be public and good API. What public and good means is up to you.

Well if it's up to me then I declare it "good" already, and we're left discussion "public". :-) I am considering simpler method names for the builder though...

> I want a working version that follows a contract where hljs provides something that lowlight can safely work with

That's the hope. :-)

> ANYTHING you expose, is public. This emitter is public. The .emitter.rootNode is public. hljs 10 broke lowlight, which was working on public APIs for years, and is now public! Public doesn’t mean it never changes.

I don't mean "public" in the purely technical sense.  Way too much in JS is public in that sense (although I'm aware there are ways to make things more private)...  I'm instating referring to "stable" and "documented" vs "beta" and "undocumented"...  sometimes I hate technical talk and if the confusion here is my fault, I'm sorry.

For example:

```js
function highlight(languageName, code, ignore_illegals, continuation) {
```

Someone calling this from v9 to v10 should get almost exactly (or exactly) the same output keys (well we added a few) - other than grammar improvements, bug fixes, or actual regressions (which we'd fix), etc.  That API is documented and very stable. I plan to deprecate it by v11, but it'll probably still continue to work past until v12... this despite the fact that probably over 50% of the internals have been rejiggered from v9 to v10.  

That's the type of documented stability/public that I'm talking about when I say stable/public. A guarantee that sites can upgrade from v10 to v10.1 to v10.2 to v10.283 and *everything* should "just work" with 0 changes by end users.

If you merely want a technically "public" API - defined as "not hidden inside a closure" then you've already got that as of v10 and I have no plans to take it away. **But I also don't consider it stable, yet.** v10.3 might break it, if it was necessary or useful for us to do so.

> You are free to create a tree! That’s what was originally asked, and what you delivered. But the original question was a tree instead of HTML. We now have both.

Yes, use whichever you prefer.  I think for many use cases this does no harm. If you can show otherwise concretely I'm all ears.

> And there are 4 ways to integrate.

LOL. I was speaking to you very technically as someone whose been forking the project for YEARS of what you could do INSTEAD of that.  If someone with no context asked me the same question "how do I use the tree" I'd say the easy/stable way was "write a builder", not give them those 4 choices I gave you. :-)

> For a tree, you could go with something that is well defined and highly used. Such as hast, but it was decided to go with a tree.

**Yes, it was an intentional choice - I wanted the parser and emitter isolated from the presentation layer.** Hast (at a quick glance) is awful close to HTML semantics, that seems to be it's key intention:

> hast is a specification for representing HTML (and embedded SVG or MathML) as an abstract syntax tree. 

It also looks a fair bit more complex than the simpler low-level solution I implemented.

> That means that anyone now has to adapt the custom hljs tree to what they want. Be it HTML or hast or ANSI or PDF. 

Well no,  you get HTML for free... :-) but... **Yes, again this is mostly intentional.**  I'm trying to build foundational layers, allow easy access for plugins and extendability - not solve (and then have to maintain!) everything directly.  Previously getting any type of tree from the parser was nigh on impossible because of the tight coupling - you decided the easiest way to do it was to fork the whole project and maintain that fork. (_I wonder if just parsing the HTML back into a tree would have been simpler._)

- **Now you can implement Hast with ~50 LOC (as an emitter it seems).** (that's pretty much what lowlight is doing, right?)
- The HTML renderer is ~50 LOC (and could possibly be simplified).
- My whole tree class + default emitter is ~80 LOC.
- Building an HTML emitter and skipping the intermediate tree would likely be ~50 LOC.
- Other simple tree-like emitters (JSON, XML, YAML) would probably have similar requirements.

> And you’re doing both a tree and HTML!

Again, this could change with time - IF it proves to be an issue in reality.

---

> I think a simpler solution would be: return whatever is returned by that __emitter. Make the emitter powerful enough for hljs HTML, lowlight hast, or ANSI, but support only 4 methods in the API. And use the emitter internally by building that HTML so it’s 100% tested.

**I agree completely - you are suggesting a simpler solution**, I just think that is at odds with the fact this is a low-level API.  I designed it purposely for integration with the parser.  **That's it's primary mission - help simplify and make the parser happy.**  If that means you have to write an `addKeyword` method that is just a more elegant abstraction around calling 3 other methods, I think that's a reasonable trade-off... 

And if we provided a subclass to extend from then (and assuming addKeyword is really always open/text/close, an assumption I question) you'd have exactly the 4 method API  you want.
You'd just extend the subclass and then not worry about those 3 extra methods if you didn't need them.  Some types of output DO require them. (like HTML, which needs to be able to close all it's tags)

> What I suggest is less public api surface than currently in v10.

Again, try writing a builder (3 methods only) and then tell me your thoughts. I'm not saying that's the best approach for `Lowlight`, but it is a simple API and for many people it'd work just fine. Many use the library to highlight a small amount of code on a website page/article.  If you want the maximum speed and ability to "bypass" default HLJS behavior, then you are naturally working at a lower-level and **I believe for a low-level API Emitter is already pretty darn simple/elegant.** (for your use case: 5 required methods, and 3 NOOPs).

> P.S. lowlight (and thus hljs) is used in many static site generators, such as Gatsby or Next.js, and it’s running on some blogs with thousands of articles: performance is really important, bazillions of lines are being highlighted

If you had a benchmark suite or server setup to measure this stuff more precisely that'd be nice.  Or even if you had a large project that you could build with say the built-in rendered plus a custom text only emitter and show numbers from both, that could prove useful to know. Right now we just use the test suite itself as a rough benchmark - since I run it all the time and would get annoyed if it were any slower.

**Comment by yyyc514** [comment]
Again: (pretty sure I said it earlier) If the tree based emitter is shown to be significantly worse/slower (a significant regression compared to a text only emitter) I think we'd definitely accept a PR to make a text only emitter the default option - if it solved the problem of how to run the test suite against both emitters. 

But so far I'm unconvinced the low-level API needs any real changes - other than perhaps a small subclass to prevent writing the NOOPs by hand. (I will think about the naming... `text` vs `addText`)... that's something I could possibly imagine changing - though it's also possible we just have different ideas on what good naming is. :)

**Comment by yyyc514** [comment]
And give me a few days to think/dream on things too... sometimes I do come around a few days later... :-)

Though you do have me debating the necessity of `toHTML()` vs a more generic `output()` or `output` getter... the assumption then being the the emitter `output()` returns whatever eventually ends up in the result object as the `value` attribute.  I think that is one of the things you mentioned above. :-)

That might remove one method from the interface IF you were happy with exposing your raw data object as an attribute on  your emitter vs a method or getter.

**Comment by yyyc514** [comment]
I think at some point you said you expected a different API for trees vs HTML, and I don't completely disagree but I think that's something we'd solve in the future with a better API overall for highlight (if we decide to make emitters a more public facing feature):

```js
var result = highlight(code, {language: "pascal", emitter: Lowlight.Emitter })
// result.value is HAST tree generated by lowlight
```

**Comment by yyyc514** [comment]
Three example cases for `addKeyword` (other than it being required by the parser):

1. A binary output format of some type... a keyword might merely be say `0x00, 'K', 0x00,"function",0x00`... no open tag, no close tag.
2. A text stream, not sure WHY LOL, but open/close wouldn't be needed at all here. Technically it might not hurt to call the NOOPs, but that's just busy work.
3. An emitter that was capturing the list of emitted commands for some other purpose (perhaps to replay them later, rewind them, re-order them, etc)... "add a keyword" is a very different concept than "new node", "add text", "close node".... and hence the emitter can preserve that fidelity if it chooses to.

I think your making some big assumptions about the output stream that I simply am not making, and hence the `addKeyword` being a requirement method - because it's a distinct concept.

Actually at a later point if all you had were nodes there would be no trivial way to distinguish a keyword from a named node...

**Comment by vladshcherbin** [comment]
It would be nice to have a function which returns just the tree which highlight function uses inside, this separates tree creation and html generation for cases when it's not needed. Something like:

```js
// returns just the tree
function buildTree(language, code) {}

// uses this function inside and producess html
function highlight(language, code) {
  const tree = buildTree(language, code)

  return { value: produceHtml(tree) }
}
```

I slightly read the above conversation and I believe this was one of wishes. I need same thing, just the tree without html stuff, a public function for doing just that.

**Comment by yyyc514** [comment]
>  I need same thing, just the tree without html stuff, a public function for doing just that.

The current behavior is intended more for advanced integrators - so I'd love to hear more about your actual use case - how and why you're using Highlight.js, etc... Often more is learned by hearing the exact problem itself than the desired solution - _because often there are much better solutions to be found_.

I'd also love to see some evidence that the HTML generation itself is actual a measurable computational expense.  _I'd wager the parsing/tree building is where MOST of the expense is._

If you ONLY need the tree **always** you could have it with a one line patch to `TokenTreeEmitter`'s prototype - simply stub/mock out `toHTML` to do no work.  Or better yet drop your own `toWhatever` function that does whatever you'd like - preferable for maximum stability it'd use `walk` - not access the tree data directly. Ie, write a Renderer (20-50 LOC for simple cases).

One other choice (if we prove that there is a real cost here) would just be to make `value` a lazy calculation instead of doing it in advance.  If you don't want the HTML, don't access `value` and you won't get it.

Simply providing your own emitter is another choice for only 30-60 LOC for simple cases.

**Comment by yyyc514** [comment]
Until I talk to more people who are using this or trying to I'm hesitant to lock down an official public API and what the "tree" is exactly...

**Comment by yyyc514** [comment]
I think it's been written before but if not my driving motivation here was to clean up the Highlight.js codebase and make it more modular - separating the output concern from the parsing concern.  Previously to do anything like this you needed to fork the whole project and then maintain that fork across time, etc.  What a pain.  Now you can build a custom emitter with only 50-60 LOC or less - using the standard distribution via NPM, because it exposes `__emitter` as a beta API.

**The end goal was never necessarily to build a `highlightAndReturnTree` function.** (that may have been the request, but it was never the goal)

The fact that we ship an internal parse tree was perhaps [or not] my getting a bit carried away. In hindsight there is some question of whether we should really just include a string emitter and then let someone else build (and maintain) a 3rd party tree emitter. That's kind of exactly what @wooorm is working on actually - just the tree - if you're happy using the standard tree format that his library provides.

_And we may still end up back with merely a string emitter in the future._ The major winning point is that now integrations are FAR easier - without requiring a fork of the library. And that if someone wanted to publish a 3rd party highlightTree() API of their own, they could trivially do so (compared to before).

**Comment by vladshcherbin** [comment]
> The current behavior is intended more for advanced integrators - so I'd love to hear more about your actual use case - how and why you're using Highlight.js, etc...

My use case is actually simple, I have a sql query and want to highlight it to show in console with colors. For console I don't need any html generation, just parsed tree to iterate and add colors.

**Comment by yyyc514** [comment]
Sounds like a perfect opportunity to me - 3rd party ANSI highlighter plugin. :-)

**Comment by wooorm** [comment]
@vladshcherbin if you're in need of something right now, than maybe https://github.com/wooorm/emphasize can help

**Comment by yyyc514** [comment]
@wooorm Figures that you already went and did it. 🥇

**Comment by vladshcherbin** [comment]
@wooorm yeah, I found it from your discussion above and already tried (it works). Thought that since there's tree generation in highlight itself now, you can probably try to drop lowlight dependency. However, I don't know the details so it's just a thought 🤔 

Thank you

**Comment by yyyc514** [comment]
Lowlight wants a very specific kind of tree. :-)  But that can be good thing for those who want to integrate with it.  I think once Lowlight switches to the new API (hopefully?) it might be the perfect place to point users to who are asking for "just give me the tree" as you were originally.

@wooorm Did you not create a separate issue for getting a new instance of HLJS, I thought you had - but I can't find it.
